Packages/com.unity.inputsystem/Documentation~/Layouts.md

```diff
@@ -295,3 +305,84 @@ const string json = @"
+
+Once registered, a precompiled layout is automatically used whenever the layout that the precompiled layout is based on is instantiated.
+
+```CSharp
+// Let's assume you have a custom device class.
+public class MyDevice : InputDevice
+{
+    // Setters for your control getters need to have at least `protected`
+    // or `internal` access so the precompiled version can use them.
+    [InputControl]
+    public ButtonControl button { get; protected set; }
+
+    // This method will *NOT* be invoked by the precompiled version. Instead, all the lookups
+    // performed here will get hardcoded into the generated C# code.
+    protected override void FinishSetup()
+    {
+        base.FinishSetup();
+
+        button = GetChildControl<ButtonControl>("button1");
+    }
+}
+
+// You register the device as a layout somewhere during startup.
+InputSystem.RegisterLayout<MyDevice>();
+
+// And you register a precompiled version of it then as well.
+InputSystem.RegisterPrecompiledLayout<PrecompiledMyDevice>(PrecompiledMyDevice.metadata);
+
+// Then the following will implicitly use the precompiled version.
+InputSystem.AddDevice<MyDevice>();
+```
+
+A precompiled layout will automatically be unregistered in the following cases:
+
+* A [layout override](#layout-overrides) is applied to one of the layouts used by the precompiled Device. This also extends to [controls](Controls.md) used by the Device.
+* A layout with the same name as one of the layouts used by the precompiled Device is registered (which replaces the layout already registered under the name).
+* A [processor](Processors.md) is registered that replaces a processor used by the precompiled Device.
+
+This causes the Input System to fall back to the non-precompiled version of the layout. Note also that a precompiled layout will not be used for layouts [derived](#layout-inheritance) from the layout the precompiled version is based on. In the example above, if someone derives a new layout from `MyDevice`, the precompiled version is unaffected (it will not be unregistered) but is also not used for the newly created type of device.
+
+```CSharp
+// Let's constinue from the example above and assume that sometime
+// later, someone replaces the builtin button with an extended version.
+InputSystem.RegisterLayout<ExtendedButtonControl>("Button");
+
+// PrecompiledMyDevice has implicitly been removed now as the ButtonControl it uses
+// has now been replaced with ExtendedButtonControl.
+```
+
+Note that you can add `#if` checks to the generated code, if needed. If you add them like in the following code snippet, the checks will be preserved even if the code is regenerated. Similarly, you can change the generated class from `public` to `internal` and the modifier will be preserved when regenerating the class.
+
+```CSharp
+//!!DEFINES
```

Comment:
@karljj1 Looking at this now, I think this is stupid. I want the user to be able to add #ifs to the generated file but required an explicit markup with `//!!DEFINES`. I think the code should simply scan for an #if and if so, preserve it.

Reply:
Removed the need for an explicit `//!!DEFINES`. Just grabs an `#if` expression now if it finds one.